src/data/setoid.lean

```diff
@@ -0,0 +1,488 @@
+            λ x y z, quotient.induction_on₃' x y z $ λ _ _ _, s.1.trans'⟩ }
+
+/-- Given an equivalence relation r on α and a map f to the quotient of α by r, an
+    equivalence relation s on the quotient induces an equivalence relation on f's domain defined
+    by x ≈ y ↔ f(x) is related to f(y) by s. -/
+def of_quotient (f : β → quotient r) (s : setoid (quotient r)) : setoid β :=
+⟨λ x y, s.rel (f x) (f y),
+ ⟨λ _, s.refl' _, λ _ _ h, s.symm' h, λ _ _ _ h1 h2, s.trans' h1 h2⟩⟩
+
+/-- Given an equivalence relation r on α, the order-preserving bijection between the set of
+    equivalence relations containing r and the equivalence relations on the quotient of α by r. -/
+def correspondence : ((≤) : {s // r ≤ s} → {s // r ≤ s} → Prop) ≃o
+  ((≤) : setoid (quotient r) → setoid (quotient r) → Prop) :=
+{ to_fun := λ s, r.to_quotient s,
+  inv_fun := λ s, subtype.mk (r.of_quotient quotient.mk' s) $
+    λ x y h, show s.rel ⟦x⟧ ⟦y⟧, from quotient.sound h ▸ s.refl' ⟦x⟧,
+  left_inv := λ s, subtype.ext.2 $ show r.of_quotient quotient.mk' (r.to_quotient s) = s.1,
+    by ext; refl,
+  right_inv := λ s, by ext; rcases a; rcases b; refl,
+  ord := λ a b, ⟨λ hle x y, quotient.induction_on₂ x y hle,
+                 λ H p q h, H ⟦p⟧ ⟦q⟧ h⟩ }
+
+-- Partitions
+
+/-- Two equivalence relations are equal iff their underlying binary operations are equal. -/
+theorem eq_iff_r_eq {r₁ r₂ : setoid α} : r₁ = r₂ ↔ r₁.rel = r₂.rel :=
+⟨λ h, h ▸ rfl, λ h, setoid.ext' $ λ x y, h ▸ iff.rfl⟩
+
+/-- If x ∈ α is in 2 elements of a set of sets partitioning α, those 2 sets are equal. -/
+lemma eq_of_mem_eqv_class {c : set (set α)}
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b')
+  {x b b'} (hc : b ∈ c) (hb : x ∈ b) (hc' : b' ∈ c) (hb' : x ∈ b') :
+  b = b' :=
+let ⟨_, _, _, h⟩ := H x in (h b hc hb).symm.trans $ h b' hc' hb'
+
+/-- Makes an equivalence relation from a set of sets partitioning α. -/
+def mk_classes (c : set (set α))
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b') :
+  setoid α :=
+⟨λ x y, ∀ b ∈ c, x ∈ b → y ∈ b, ⟨λ _ _ _ hx, hx,
+ λ x _ h _ hb hy, let ⟨z, hc, hx, hz⟩ := H x in
+    eq_of_mem_eqv_class H hc (h z hc hx) hb hy ▸ hx,
+ λ x y z h1 h2 b hc hb, let ⟨v, hvc, hy, hv⟩ := H y in let ⟨w, hwc, hz, hw⟩ := H z in
+    (eq_of_mem_eqv_class H hwc hz hvc $ h2 v hvc hy).trans
+      (eq_of_mem_eqv_class H hvc hy hc $ h1 b hc hb) ▸ hz⟩⟩
+
+/-- Makes the equivalence classes of an equivalence relation. -/
+def classes (r : setoid α) : set (set α) :=
+{s | ∃ y, s = {x | r.rel x y}}
+
+lemma mem_classes (r : setoid α) (y) : {x | r.rel x y} ∈ r.classes := ⟨y, rfl⟩
+
+/-- Two equivalence relations are equal iff all their equivalence classes are equal. -/
+lemma eq_iff_classes_eq' {r₁ r₂ : setoid α} :
+  r₁ = r₂ ↔ ∀ x, {y | r₁.rel x y} = {y | r₂.rel x y} :=
+⟨λ h x, h ▸ rfl, λ h, ext' $ λ x, (set.ext_iff _ _).1 $ h x⟩
+
+/-- Two equivalence relations are equal iff their equivalence classes are equal. -/
+lemma classes_inj {r₁ r₂ : setoid α} :
```

Comment:
I guess this should be called `eq_iff_classes_eq` (which explains the `'` on the name of the previous lemma?)

Reply:
Thanks for the other comments, I'm applying them at the moment; Chris suggested ```classes_inj``` instead of ```eq_iff_classes_eq```. (Should I avoid resolving earlier conversations? I did it to make unresolved ones easier to find.)